File: client/src/main/java/dev/throwouterror/game/client/mesh/CubeMesh.java
/*
 * Copyright (c) Creepinson
 */

package dev.throwouterror.game.client.mesh;

import org.joml.Vector3f;

/**
 * @author Theo Paris
 */
public class CubeMesh extends Mesh {
    public CubeMesh() {
        super(
                new Vector3f(1.0f, 1.0f, -1.0f), // Top Right Of The Quad (Top)
                new Vector3f(-1.0f, 1.0f, -1.0f), // Top Left Of The Quad (Top)
                new Vector3f(-1.0f, 1.0f, 1.0f), // Bottom Left Of The Quad (Top)
                new Vector3f(1.0f, 1.0f, 1.0f), // Bottom Right Of The Quad (Top)

                new Vector3f(1.0f, -1.0f, 1.0f), // Top Right Of The Quad
                new Vector3f(-1.0f, -1.0f, 1.0f), // Top Left Of The Quad
                new Vector3f(-1.0f, -1.0f, -1.0f), // Bottom Left Of The Quad
                new Vector3f(1.0f, -1.0f, -1.0f),// Bottom Right Of The Quad

                new Vector3f(1.0f, 1.0f, 1.0f),// Top Right Of The Quad (Front)
                new Vector3f(-1.0f, 1.0f, 1.0f),// Top Left Of The Quad (Front)
                new Vector3f(-1.0f, -1.0f, 1.0f),// Bottom Left Of The Quad
                new Vector3f(1.0f, -1.0f, 1.0f),// Bottom Right Of The Quad

                new Vector3f(1.0f, -1.0f, -1.0f), // Bottom Left Of The Quad
                new Vector3f(-1.0f, -1.0f, -1.0f), // Bottom Right Of The Quad
                new Vector3f(-1.0f, 1.0f, -1.0f), // Top Right Of The Quad (Back)
                new Vector3f(1.0f, 1.0f, -1.0f), // Top Left Of The Quad (Back)

                new Vector3f(-1.0f, 1.0f, 1.0f),// Top Right Of The Quad (Left)
                new Vector3f(-1.0f, 1.0f, -1.0f), // Top Left Of The Quad (Left)
                new Vector3f(-1.0f, -1.0f, -1.0f), // Bottom Left Of The Quad
                new Vector3f(-1.0f, -1.0f, 1.0f), // Bottom Right Of The Quad

                new Vector3f(1.0f, 1.0f, -1.0f), // Top Right Of The Quad (Right)
                new Vector3f(1.0f, 1.0f, 1.0f), // Top Left Of The Quad
                new Vector3f(1.0f, -1.0f, 1.0f),// Bottom Left Of The Quad
                new Vector3f(1.0f, -1.0f, -1.0f) // Bottom Right Of The Quad
        );
    }
}

File: client/src/main/java/dev/throwouterror/game/client/engine/mesh/CubeMesh.kt
package dev.throwouterror.game.client.engine.mesh

import dev.throwouterror.util.math.Tensor

/**
 * @author Theo Paris
 */
class CubeMesh : Mesh() {
    init {
        // Front
        vertices.add(Tensor(.5f, .5f, .5f))
        vertices.add(Tensor(.5f, .5f, .5f))
        vertices.add(Tensor(-.5f, .5f, .5f))
        vertices.add(Tensor(-.5f, -.5f, .5f))
        vertices.add(Tensor(.5f, .5f, .5f))
        vertices.add(Tensor(-.5f, -.5f, .5f))
        vertices.add(Tensor(.5f, -.5f, .5f))
        // Back
        vertices.add(Tensor(-.5f, .5f, -.5f))
        vertices.add(Tensor(.5f, .5f, -.5f))
        vertices.add(Tensor(.5f, -.5f, -.5f))
        vertices.add(Tensor(-.5f, .5f, -.5f))
        vertices.add(Tensor(.5f, -.5f, -.5f))
        vertices.add(Tensor(-.5f, -.5f, -.5f))
        // Left
        vertices.add(Tensor(-.5f, .5f, .5f))
        vertices.add(Tensor(-.5f, .5f, -.5f))
        vertices.add(Tensor(-.5f, -.5f, -.5f))
        vertices.add(Tensor(-.5f, .5f, .5f))
        vertices.add(Tensor(-.5f, -.5f, -.5f))
        vertices.add(Tensor(-.5f, -.5f, .5f))
        // Right
        vertices.add(Tensor(.5f, .5f, -.5f))
        vertices.add(Tensor(.5f, .5f, .5f))
        vertices.add(Tensor(.5f, -.5f, .5f))
        vertices.add(Tensor(.5f, .5f, -.5f))
        vertices.add(Tensor(.5f, -.5f, .5f))
        vertices.add(Tensor(.5f, -.5f, -.5f))
        // Top
        vertices.add(Tensor(.5f, .5f, -.5f))
        vertices.add(Tensor(-.5f, .5f, -.5f))
        vertices.add(Tensor(-.5f, .5f, .5f))
        vertices.add(Tensor(.5f, .5f, -.5f))
        vertices.add(Tensor(-.5f, .5f, .5f))
        vertices.add(Tensor(.5f, .5f, .5f))
        // Bottom
        vertices.add(Tensor(-.5f, -.5f, -.5f))
        vertices.add(Tensor(.5f, -.5f, -.5f))
        vertices.add(Tensor(.5f, -.5f, .5f))
        vertices.add(Tensor(-.5f, -.5f, -.5f))
        vertices.add(Tensor(.5f, -.5f, .5f))
        vertices.add(Tensor(-.5f, -.5f, .5f))

        // Indices
        // front
        indices.addAll(arrayListOf(
                0,
                1,
                2,
                2,
                3,
                0,
                // right
                1,
                5,
                6,
                6,
                2,
                1,
                // back
                7,
                6,
                5,
                5,
                4,
                7,
                // left
                4,
                0,
                3,
                3,
                7,
                4,
                // bottom
                4,
                5,
                1,
                1,
                0,
                4,
                // top
                3,
                2,
                6,
                6,
                7,
                3
        ))
    }
}